onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java

```diff
@@ -111,33 +111,34 @@ public TaskContextImpl(int startId) {
         public void taskComplete() {
             //Log.d(TAG, "Task complete: " + mStartId);
             // If we have notifications, then we can't stop ourselves.
-            if (mNotifications.isEmpty()) {
+            //TODO: Is there actually a reason to not stop ourselves if we are no longer storing notifications?
```

Comment:
@cagryInside Could you please take a look at this code?  It's been a while since I've dived into this, and I'd like another pair of eyes on it as well.  

`TripService` acts as a container service for a thread pool.  The thread pool can contain executing threads for various tasks related to the reminders feature - the `SchedulerTask` (executed when a new reminder is saved, or after boot (via `BootstrapService`) to register all reminders saved in the database with the Android platform), the `PollerTask` (triggered by the platform via `AlarmReceiver` to begin polling the server 30 min ahead of the scheduled time of the arrival for which the reminder is set), the `NotifierTask` (triggered by OBA Android when we should fire a notification for a reminder), or the `CancelNotifyTask` (for when a notification is canceled).

My gut feeling is that we shouldn't be stopping this service every time this executes - IIRC this was necessary to keep querying the server in the background to ensure we remind the user (I believe in the `PollerTask`) when some of the other tasks were executing.

Previously we were holding a reference to all notifications in `mNotifications`, which is (I believe) why this check existed, and I believe this prevented the Service from shutting down if other tasks (e.g., `SchedulerTask`) were still running in Threads in the same service.  My feeling is that we should be checking to make sure that no other Threads are still executing before we shut down the service.

I think the best way to test this is to save a reminder (Reminder A).  Reminder A will trigger the `PollerTask` to start executing 30 min ahead of the scheduled arrival time.  While Reminder A's `PollerTask` is executing, schedule a new reminder (Reminder B), which will start a  `SchedulerTask` in the same service.  Reminder A's `PollerTask` should still be executing after the `SchedulerTask` for Reminder B completes.

We should also test this when turning the device off and back on while a `PollerTask` is executing, to make sure everything gets re-scheduled.

@charles-b-stb Could you please give this a test run as well?